include/aspect/particle/property/cpo.h

```diff
@@ -0,0 +1,361 @@
+                                const std::pair<std::vector<double>, std::vector<Tensor<2,3> > > &derivatives,
+                                const double strain_rate_second_invariant,
+                                const double dt) const;
+
+          /**
+           * Updates the volume fractions and rotation matrices with a Crank-Nicolson scheme.
+           */
+          double
+          advect_Crank_Nicolson(std::vector<double> &volume_fractions,
+                                std::vector<Tensor<2,3> > &a_cosine_matrices,
+                                const std::pair<std::vector<double>, std::vector<Tensor<2,3> > > &derivatives,
+                                std::vector<double> &previous_volume_fraction_derivatives,
+                                std::vector<Tensor<2,3> > &previous_a_cosine_matrices_derivatives,
+                                const double strain_rate_second_invariant,
+                                const double dt) const;
+
+
+          /**
+           * A interface function which delegates the computation the derivatives
+           * to specialized functions.
+           */
+          std::pair<std::vector<double>, std::vector<Tensor<2,3> > >
+          compute_derivatives(const std::vector<double> &volume_fractions,
+                              const std::vector<Tensor<2,3> > &a_cosine_matrices,
+                              const SymmetricTensor<2,3> &strain_rate_nondimensional,
+                              const Tensor<2,3> &velocity_gradient_tensor_nondimensional,
+                              const double volume_fraction_mineral,
+                              const std::array<double,4> &ref_resolved_shear_stress) const;
+
+          /**
+           * Computes the update for the property rotating passively with the particle.
+           */
+          std::pair<std::vector<double>, std::vector<Tensor<2,3> > >
+          compute_derivatives_spin_tensor(const std::vector<double> &volume_fractions,
+                                          const std::vector<Tensor<2,3> > &a_cosine_matrices,
+                                          const SymmetricTensor<2,3> &strain_rate_nondimensional,
+                                          const Tensor<2,3> &velocity_gradient_tensor_nondimensional,
+                                          const double volume_fraction_mineral,
+                                          const std::array<double,4> &ref_resolved_shear_stress) const;
+
+
+          /**
+           * Declare the parameters this class takes through input files.
+           */
+          static
+          void
+          declare_parameters (ParameterHandler &prm);
+
+          /**
+           * Read the parameters this class declares from the parameter file.
+           */
+          void
+          parse_parameters (ParameterHandler &prm) override;
+
+          /**
+           * Return the number of grains per particle
+           */
+          static
```

Comment:
We try to avoid static functions whenever possible. Is this necessary? It may create problems once you want to change the number of grains or minerals based on input files.

Reply:
> We try to avoid static functions whenever possible. Is this necessary?

I agree that it is not ideal, but part of the reason why they are static is because it is information I would like to directly use in particle property plugins and static particle property functions (such as the store and load functions, which are use by these other plugins) which directly builds on this information.

> It may create problems once you want to change the number of grains or minerals based on input files.

With the current code you can already set the number of minerals and grains per mineral in the input file, so I am not sure what you mean.